### ISSUE TITLE
[HTTP/3] Stress QUIC mangled data

### ISSUE THREAD

**Issue by ManickaP** [dotnet/runtime#72739]
While investigating `System.Net.Http.HttpRequestException: The server returned an invalid or unrecognized response.` and similar errors with H/3 stress I noticed that data sent from the server do not correspond to the data received on the client. Specifically, the first 2 bytes get mangled:
```
// Data from server from StreamSend
2022-07-24 15:33:11.189 +02:00 [DBG] 13:33:11.1891745[Info] thisOrContextObject: QuicStream#28784787, memberName: WriteAsync, message: [strm][0x7FB794029990] SEND 012C0000D9561D53756E2C203234204A756C20323032322031333A33333A313120474D545F4D074B65737472656C00417F000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F202122232425262728292A2B2C2D2E2F303132333435363738393A3B3C3D3E3F404142434445464748494A4B4C4D4E4F505152535455565758595A5B5C5D5E5F606162636465666768696A6B6C6D6E6F707172737475767778797A7B7C7D7E7F000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F202122232425262728292A2B2C2D2E2F303132333435363738393A3B3C3D3E3F404142434445464748494A4B4C4D4E4F505152535455565758595A5B5C5D5E5F606162636465666768696A6B6C6D6E6F707172737475767778797A7B7C7D7E7F000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F202122232425262728292A2B2C2D2E2F303132333435363738393A3B3C3D3E3F404142434445464748494A4B4C4D4E4F505152535455565758595A5B5C5D5E5F606162636465666768696A6B6C6D6E6F707172737475767778797A7B7C7D7E
// Data from client from RECEIVE event
13:33:11.1977316[Info] thisOrContextObject: QuicStream#27524287, memberName: HandleEventReceive, message:                           [strm][0x7FB7E801A860] RECEIVE 422C0000D9561D53756E2C203234204A756C20323032322031333A33333A313120474D545F4D074B65737472656C00417F000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F202122232425262728292A2B2C2D2E2F303132333435363738393A3B3C3D3E3F404142434445464748494A4B4C4D4E4F505152535455565758595A5B5C5D5E5F606162636465666768696A6B6C6D6E6F707172737475767778797A7B7C7D7E7F000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F202122232425262728292A2B2C2D2E2F303132333435363738393A3B3C3D3E3F404142434445464748494A4B4C4D4E4F505152535455565758595A5B5C5D5E5F606162636465666768696A6B6C6D6E6F707172737475767778797A7B7C7D7E7F000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F202122232425262728292A2B2C2D2E2F303132333435363738393A3B3C3D3E3F404142434445464748494A4B4C4D4E4F505152535455565758595A5B5C5D5E5F606162636465666768696A6B6C6D6E6F707172737475767778797A7B7C7D7E

// Server
2022-07-24 16:22:07.122 +02:00 [DBG] 14:22:07.1225155[Info] thisOrContextObject: QuicStream#64247531, memberName: WriteAsync, message: [strm][0x7F6BD401BAC0] SEND    012C0000D9561D53756E2C203234204A756C20323032322031343A32323A303620474D545F4D074B65737472656C0040C4000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F202122232425262728292A2B2C2D2E2F303132333435363738393A3B3C3D3E3F404142434445464748494A4B4C4D4E4F505152535455565758595A5B5C5D5E5F606162636465666768696A6B6C6D6E6F707172737475767778797A7B7C7D7E7F000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F202122232425262728292A2B2C2D2E2F303132333435363738393A3B3C3D3E3F40414243
// Client
14:22:07.1297277[Info] thisOrContextObject: QuicStream#19014160, memberName: HandleEventReceive, message:                              [strm][0x7F6B840233B0] RECEIVE 6B6B0000D9561D53756E2C203234204A756C20323032322031343A32323A303620474D545F4D074B65737472656C0040C4000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F202122232425262728292A2B2C2D2E2F303132333435363738393A3B3C3D3E3F404142434445464748494A4B4C4D4E4F505152535455565758595A5B5C5D5E5F606162636465666768696A6B6C6D6E6F707172737475767778797A7B7C7D7E7F000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F202122232425262728292A2B2C2D2E2F303132333435363738393A3B3C3D3E3F40414243

```

Happens only with "GET Aborted" as far as I can tell, fairly regularly.

**Comment by ManickaP** [comment]
```C#
 endpoints.MapGet("/abort", async context =>
{
    // Server writes some content, then aborts the connection
    string content = CreateResponseContent(context);
    await context.Response.WriteAsync(content.Substring(0, content.Length / 2));
    context.Abort();
});
```

Root cause of the problem is that we unblock `WriteAsync` as part of `Abort` which lets the caller to reuse the memory passed to MsQuic. This is possible to happen in parallel on Kestrel due indirection between app calling `Reponse.WriteAsync` and the actual transport doing `QuicStream.WriteAsync` since there is pipeline in between.

So we either need to wait for `SEND_COMPLETE` or do our own buffer copy.

**Comment by ManickaP** [comment]
Mulling this over, I think we need to do a copy. In case of a cancellation, we also unblock the waiter before we get `SEND_COMPLETE` which might overwrite data in the buffer passed to MsQuic. And we certainly cannot wait for the event in case of a cancellation, that would made it useless.

**Comment by wfurt** [comment]
If new cancel/abort, _can_ we wait for notification from MsQuic? It feels like that should not be too long as all should be processable locally. If we copy, we will alway pay penalty for (hopefully) rare case.

**Comment by ManickaP** [comment]
It nullifies the cancel, since we're basically waiting for send confirmation (it should be fast for internal buffering but might block if msquic has full buffers). As a result cancel wouldn't do anything, we might just ignore it.

**Comment by CarnaViire** [comment]
> If new cancel/abort, can we wait for notification from MsQuic?

> It nullifies the cancel, since we're basically waiting for send confirmation (it should be fast for internal buffering but might block if msquic has full buffers). As a result cancel wouldn't do anything, we might just ignore it.

I believe MsQuic would stop waiting and it would deliver the event SEND_COMPLETED(Canceled=1) as soon as it can... at least that's my understanding of the code https://github.com/microsoft/msquic/blob/41a42f0fbf51f217a6bee11e8955a8f1ac179b38/src/core/stream_send.c#L159-L165

**Comment by wfurt** [comment]
I would expect there would be some way how to abort the send... perhaps the one @CarnaViire pointed out. cc: @nibanks for confirmation.

**Comment by ManickaP** [comment]
I recollected the server logs and extracted the case when it ends up with the mangled data on the client:
```
   4846: 2022-07-26 09:30:39.863 +02:00 [DBG] 07:30:39.8638102[Info] thisOrContextObject: MsQuicContextSafeHandle#26523563, memberName: .ctor, message: [strm][0x7F9030068FC0] MsQuicSafeHandle created
   4848: 2022-07-26 09:30:39.863 +02:00 [DBG] 07:30:39.8638539[Info] thisOrContextObject: MsQuicContextSafeHandle#26523563, memberName: .ctor, message: [strm][0x7F9030068FC0] [conn][0x7F9060001830] ref count incremented
   4851: 2022-07-26 09:30:39.863 +02:00 [DBG] Stream id "0HMJER2ULGIHQ:00000204" type "Bidirectional" accepted.
   4856: 2022-07-26 09:30:39.864 +02:00 [DBG] 07:30:39.8640184[Info] thisOrContextObject: QuicStream#2539263, memberName: ReadAsync, message: [strm][0x7F9030068FC0] Stream reading into memory of '4096' bytes.
   4939: 2022-07-26 09:30:39.866 +02:00 [DBG] 07:30:39.8661158[Info] thisOrContextObject: QuicStream#2539263, memberName: NativeCallback, message: [strm][0x7F9030068FC0] Received event RECEIVE
   4941: 2022-07-26 09:30:39.866 +02:00 [DBG] 07:30:39.8661744[Info] thisOrContextObject: QuicStream#2539263, memberName: NativeCallback, message: [strm][0x7F9030068FC0] Received event PEER_SEND_SHUTDOWN
   4945: 2022-07-26 09:30:39.866 +02:00 [DBG] Trace id "0HMJER2ULGIHQ:00000204": HTTP/3 stream error "H3_INTERNAL_ERROR". An abort is being sent to the stream.
   4950: 2022-07-26 09:30:39.866 +02:00 [DBG] Stream id "0HMJER2ULGIHQ:00000204" aborted by application with error code 258 because: "The connection was aborted by the application.".
   4985: 2022-07-26 09:30:39.867 +02:00 [DBG] Stream id "0HMJER2ULGIHQ:00000204" shutting down writes because: "The connection was aborted by the application.".
   4987: 2022-07-26 09:30:39.867 +02:00 [DBG] 07:30:39.8670404[Info] thisOrContextObject: QuicStream#2539263, memberName: WriteAsync, message: [strm][0x7F9030068FC0] Stream writing memory of '60' bytes while not completing writes.
   5085: 2022-07-26 09:30:39.870 +02:00 [DBG] 07:30:39.8707855[Info] thisOrContextObject: QuicStream#2539263, memberName: NativeCallback, message: [strm][0x7F9030068FC0] Received event SEND_COMPLETE
   5093: 2022-07-26 09:30:39.870 +02:00 [DBG] 07:30:39.8708440[Info] thisOrContextObject: QuicStream#2539263, memberName: HandleEventSendComplete, message: [strm][0x7F9030068FC0] Received event SEND_COMPLETE Canceled = 0
   5298: 2022-07-26 09:30:39.877 +02:00 [DBG] 07:30:39.8776561[Info] thisOrContextObject: QuicStream#2539263, memberName: NativeCallback, message: [strm][0x7F9030068FC0] Received event SEND_SHUTDOWN_COMPLETE
   5305: 2022-07-26 09:30:39.877 +02:00 [DBG] 07:30:39.8777093[Info] thisOrContextObject: QuicStream#2539263, memberName: NativeCallback, message: [strm][0x7F9030068FC0] Received event SHUTDOWN_COMPLETE
   5307: 2022-07-26 09:30:39.877 +02:00 [DBG] 07:30:39.8779245[Info] thisOrContextObject: MsQuicContextSafeHandle#26523563, memberName: ReleaseHandle, message: [strm][0x7F9030068FC0] MsQuicSafeHandle released
   5315: 2022-07-26 09:30:39.877 +02:00 [DBG] 07:30:39.8779787[Info] thisOrContextObject: MsQuicContextSafeHandle#26523563, memberName: ReleaseHandle, message: [strm][0x7F9030068FC0] [conn][0x7F9060001830] ref count decremented
```

It takes about 3ms and the SEND_COMPLETE is not canceled.
You can also see the line numbers to gauge how many other events is happening in between.

**Comment by ManickaP** [comment]
We could, with a bit more time investment, do our own data buffering and not use MsQuic one. But we should not do that with just one buffer based on the recommendation here: https://github.com/microsoft/msquic/blob/main/docs/Streams.md#send-buffering
> To fill the pipe in this mode, the app is responsible for keeping enough sends pending at all times to ensure the connection doesn't go idle. MsQuic indicates the amount of data the app should keep pending in the QUIC_STREAM_EVENT_IDEAL_SEND_BUFFER_SIZE event. The app should always have at least two sends pending at a time. If only a single send is used, the connection can go idle for the time between that send is completed and the new send is queued.

**The app should always have at least two sends pending at a time.**

**Comment by nibanks** [comment]
> If new cancel/abort, can we wait for notification from MsQuic?

If you're asking "can you cancel a send" you can't cancel an individual send, but you can abort (shutdown) the send direction of your stream, which should give you a send completion (cancelled) pretty immediately.

**Comment by wfurt** [comment]
Is there option to abort without notifying the peer? (and possibly abort with particular error code later)

**Comment by ManickaP** [comment]
> If you're asking "can you cancel a send" you can't cancel an individual send, but you can abort (shutdown) the send direction of your stream, which should give you a send completion (cancelled) pretty immediately.

Not what I've seen in the logs above, we'd still get SEND_COMPLETE without Canceled and than `SEND_SHUTDOWN_COMPLETE`.

**Comment by wfurt** [comment]
may be race condition? To me the important part is that we can do that without waiting for round-trip to peer.

**Comment by CarnaViire** [comment]
#72746 is in, so the mangled data should disappear; keeping the issue open for the possible fix without copying.

**Comment by ManickaP** [comment]
@CarnaViire could you please file an for the follow up with doing our own send buffering (unless we already have some)? So that we don't have to copy the data twice (us + msquic). I think that should be addressed early in 8.0.

**Comment by karelz** [comment]
I would suggest to close this issue -- as in "the problem is gone" and if it reappears we have a problem.
We should file separate issue as perf improvement to make better fix in 8.0.

**Comment by CarnaViire** [comment]
Closing as fixed by https://github.com/dotnet/runtime/pull/72746

**Comment by CarnaViire** [comment]
> could you please file an for the follow up with doing our own send buffering (unless we already have some)? So that we don't have to copy the data twice (us + msquic). I think that should be addressed early in 8.0.

I didn't find one so I've filed https://github.com/dotnet/runtime/issues/73691
